### PR TITLE
get token balances after transaction in simulation

### PR DESCRIPTION
**Pull request by KillariDev**
Backend work for https://github.com/DarkFlorist/TheInterceptor/issues/209

image

### CODE REVIEW

File: extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts [hunk 1]
```diff
@@ -663,4 +692,76 @@ export class SimulationModeEthereumClientService {
...
+	) => {
+		const tokenBalancesAfter: TokenBalancesAfter[] = []
+		for (let resultIndex = 0; resultIndex < multicallResult.length; resultIndex++) {
+			const singleResult = multicallResult[resultIndex]
+			const balances = await SimulationModeEthereumClientService.getTokenBalances(
+				ethereumClientService,
+				signedTxs.slice(0, resultIndex),
+				SimulationModeEthereumClientService.getAddressesInteractedWithERC20s(singleResult.statusCode === 'success' ? singleResult.events : []),
+				blockNumber
+			)
```
Comment (MicahZoltu) [inline extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts:762]: This code looks almost identical to the code above, would there be valuae in extracting out a helper function?
```suggestion
			const transactionStack = signedTxs.slice(0, resultIndex)
			const eventsInTopTransaction = SimulationModeEthereumClientService.getAddressesInteractedWithERC20s(singleResult.statusCode === 'success' ? singleResult.events : [])
			const balances = await SimulationModeEthereumClientService.getTokenBalances(ethereumClientService, transactionStack, eventsInTopTransaction, blockNumber)
```

File: extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts [hunk 2]
```diff
@@ -663,4 +692,76 @@ export class SimulationModeEthereumClientService {
...
+		}
+		return tokenOwners
+	}
+
+	static readonly getTokenBalancesAfter = async (
+		ethereumClientService: EthereumClientService,
+		signedTxs: EthereumSignedTransaction[] = [],
+		multicallResult: MulticallResponse,
+		blockNumber: bigint,
+	) => {
```
Comment (MicahZoltu) [inline extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts:753]: Multiline doesn't seem to be adding value here.
```suggestion
	static readonly getTokenBalancesAfter = async (ethereumClientService: EthereumClientService, signedTxs: EthereumSignedTransaction[] = [], multicallResult: MulticallResponse, blockNumber: bigint) => {
```

**Review by MicahZoltu — APPROVED** [review]
A bunch of non-blocking style suggestions.

File: extension/app/ts/background/settings.ts [hunk 3]
```diff
@@ -156,18 +156,22 @@ export async function setPendingInterceptorAccessRequestPromise(promise: Pending
...
+		simulationId: 0,
+		simulationState: undefined,
+		visualizerResults: undefined,
+		addressBookEntries: [],
+		tokenPrices: [],
+		activeAddress: undefined
+	}
+	try {
+		const parsed = funtypes.Union(funtypes.Undefined, SimulationResults).parse(results)
+		if (parsed === undefined) return emptyResults
```
[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/background/settings.ts:169]: hmm, I think its better to keep default settings in code instead of in typing 🤔

**Pull request merged**